clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java

```diff
@@ -0,0 +1,595 @@
+        this.defaultOffsetCommitCallback = defaultOffsetCommitCallback;
+        this.autoCommitEnabled = autoCommitEnabled;
+
+        this.protocolMap = new HashMap<>();
+        for (PartitionAssignor assignor : assignors)
+            this.protocolMap.put(assignor.name(), assignor);
+
+        addMetadataListener();
+
+        if (autoCommitEnabled)
+            scheduleAutoCommitTask(autoCommitIntervalMs);
+
+        this.sensors = new ConsumerCoordinatorMetrics(metrics, metricGrpPrefix, metricTags);
+    }
+
+    @Override
+    public String protocolType() {
+        return "consumer";
+    }
+
+    @Override
+    public LinkedHashMap<String, ByteBuffer> metadata() {
+        LinkedHashMap<String, ByteBuffer> metadata = new LinkedHashMap<>();
+        for (PartitionAssignor assignor : protocolMap.values()) {
+            Subscription subscription = assignor.subscription(subscriptions.subscription());
+            metadata.put(assignor.name(), ConsumerProtocol.serializeSubscription(subscription));
+        }
+        return metadata;
+    }
+
+    private void addMetadataListener() {
+        this.metadata.addListener(new Metadata.Listener() {
+            @Override
+            public void onMetadataUpdate(Cluster cluster) {
+                if (subscriptions.hasPatternSubscription()) {
+                    final List<String> topicsToSubscribe = new ArrayList<>();
+
+                    for (String topic : cluster.topics())
+                        if (subscriptions.getSubscribedPattern().matcher(topic).matches())
+                            topicsToSubscribe.add(topic);
+
+                    subscriptions.changeSubscription(topicsToSubscribe);
+                    metadata.setTopics(subscriptions.groupSubscription());
+                }
+
+                // check if there are any changes to the metadata which should trigger a rebalance
+                if (metadataSnapshot.update(subscriptions, cluster) && subscriptions.partitionsAutoAssigned())
+                    subscriptions.needReassignment();
+            }
+        });
+    }
+
+    @Override
+    protected void onJoin(int generation,
+                          String memberId,
+                          String assignmentStrategy,
+                          ByteBuffer assignmentBuffer) {
+        PartitionAssignor assignor = protocolMap.get(assignmentStrategy);
+        if (assignor == null)
+            throw new IllegalStateException("Coordinator selected invalid assignment protocol: " + assignmentStrategy);
+
+        Assignment assignment = ConsumerProtocol.deserializeAssignment(assignmentBuffer);
+
+        // set the flag to refresh last committed offsets
+        subscriptions.needRefreshCommits();
+
+        // update partition assignment
+        subscriptions.changePartitionAssignment(assignment.partitions());
+
+        // give the assignor a chance to update internal state based on the received assignment
+        assignor.onAssignment(assignment);
```

Comment:
The idea is to have the local assignment returned in onAssignment included in the user data of the next subscription. All subscriptions are forwarded to the leader, so it would be able to use the previous assignments found in the member's subscription userData to compute the next assignments.